ApeWorX/ape
# test: allow CLI test projects to opt-out of tests

### What I did

Make it so that the projects in `tests.integration.cli` can have a `tests-config.yaml` file that includes a list of `skip-tests`. Any test in the skip-tests data will not run for that project!

Additionally, I refactored some of the integration tests to use this new approach instead of having early returns and stuff.

This _should_ help us all add integration tests... and make our lives easier.

### How I did it

The hook `pytest_collection_modifyitem()` and filtering.

Also, reading the `tests-config.yaml` file for each project and all the tests in `tests.integration.cli`

It still uses the parametrized fixture approach. See the note below for more historical details!

___

I started off working with that branch that @fubuloubu and I started on together during a peer-programming session.
On that branch, we were fiddling with creating custom test items from a parent node that is the project's `tests-config.yaml` file. I got something almost working, but it required copying a lot of internal code from `pytest` and it was not working out that great. I then learned of `pytest_collection_modifyitems()` and how it can be used to filter out tests. This approach has turned out to give us exactly what we need without tweaking the internals of `pytest` too much.. I structured the code in this PR to be re-usable for if we do end up needing another approach.

Also, with what I have learned about creating custom `pytest` nodes, I think that approach could still make sense for `vyper` or elsewhere.

### How to verify it

The tests should run!  If you copy only the test changes to the main branch, they fail because the `if` checks are missing and tests are configured to run per project elsewhere (in the `tests-config.yaml` file).
Also try messing with the `tests-config.yaml`!

### Checklist
